File: py/src/tradekit/fyers/transport.py
"""Throttled, retrying HTTP transport for the FYERS API v3.

Mirrors ``go/fyers/{errors,limiter,transport}.go``. The token bucket and the
idempotency rule are the Upstox transport's: order mutations are never retried,
because retrying a POST that placed an order places a second one.

What differs from Upstox is the envelope. Every FYERS response carries
``s`` (``ok``/``error``), ``code`` and ``message``, and a refusal can arrive
under HTTP 200 -- an order rejection is ``{"s":"error","code":-99}`` with a
200 status -- so the body is inspected on every response, not only on 4xx.
"""

from __future__ import annotations

import json
import logging
import threading
import time
from collections.abc import Callable
from typing import Any

import httpx

__all__ = [
    "APIError",
    "RateLimitedError",
    "RejectedError",
    "TokenBucket",
    "TokenExpiredError",
    "TransientError",
    "Transport",
]

log = logging.getLogger(__name__)

CODE_TOKEN_EXPIRED = -8
CODE_TOKEN_INVALID = -15
CODE_TOKEN_UNVERIFIABLE = -16
CODE_TOKEN_BAD = -17
CODE_RATE_LIMITED = -429
"""FYERS's application-level codes that change how a refusal is handled.

The token codes matter because FYERS answers an expired token with HTTP 401 on
most endpoints but not all; the body's code is the reliable signal. The
rate-limit code matters for the opposite reason: a rate limit read as a
rejection would stop a sync that only needed to wait.
"""

TOKEN_CODES = frozenset({CODE_TOKEN_EXPIRED, CODE_TOKEN_INVALID, CODE_TOKEN_UNVERIFIABLE, CODE_TOKEN_BAD})

DEFAULT_ATTEMPTS = 4
DEFAULT_RETRY_BASE = 2.0
MAX_RETRY_WAIT = 60.0


class APIError(RuntimeError):
    """A refusal from FYERS, carrying enough of the response to diagnose it."""

    def __init__(self, status: int, code: int, message: str) -> None:
        """Record the status, FYERS's own code, and its description."""
        detail = f"code {code}: {message}" if code else message
        super().__init__(f"fyers: HTTP {status}: {detail}")
        self.status = status
        self.code = code
        self.message = message

    @property
    def retryable(self) -> bool:
        """Whether repeating the call could succeed."""
        return False


class TokenExpiredError(APIError):
    """The access token is no longer valid; the caller must log in again.

    Distinct from :class:`TransientError` because retrying with the same token
    cannot succeed. This is the one refusal whose remedy is a login.
    """


class TransientError(APIError):
    """Transport, FYERS's infrastructure, or a rate limit. Worth repeating."""

    @property
    def retryable(self) -> bool:
        """Always true: that is what makes this class the transient one."""
        return True


class RateLimitedError(TransientError):
    """FYERS saying "ask again later", by status or by envelope code."""


class RejectedError(APIError):
    """FYERS refused the request on its merits.

    Repeating it unchanged will be refused again, so it is deliberately not
    retryable. An unrecognised refusal lands here rather than in
    :class:`TransientError`: the cost of not retrying a transient failure is one
    missed call, and the cost of retrying a rejection is the same bad order sent
    repeatedly.
    """


class TokenBucket:
    """Thread-safe client-side throttle.

    A burst of ``capacity`` then a steady ``rate`` per second. It cannot prevent
    a rate limit on a bulk job -- FYERS's per-minute window is longer than any
    burst -- only keep the burst from provoking one immediately. That matters
    more here than at Upstox: three per-minute breaches block the account for
    the rest of the day.
    """

    def __init__(self, rate: float, capacity: float, sleep: Callable[[float], None] = time.sleep) -> None:
        """Initialise the bucket. A non-positive rate disables throttling."""
        self.rate = rate
        self.capacity = capacity
        self._tokens = capacity
        self._updated = time.monotonic()
        self._lock = threading.Lock()
        self._sleep = sleep

    def acquire(self, tokens: float = 1.0) -> None:
        """Block until ``tokens`` are available, then consume them."""
        if self.rate <= 0:
            return
        while True:
            with self._lock:
                now = time.monotonic()
                self._tokens = min(self.capacity, self._tokens + (now - self._updated) * self.rate)
                self._updated = now
                if self._tokens >= tokens:
                    self._tokens -= tokens
                    return
                wait = (tokens - self._tokens) / self.rate
            self._sleep(min(wait, 1.0))


def _envelope(body: str) -> tuple[str, int, str]:
    """Extract ``s``, ``code`` and ``message`` from a response body.

    A body that is not a JSON object at all yields an empty envelope; the
    history endpoint on an empty range has been seen to answer with a bare
    array.
    """
    try:
        payload = json.loads(body)
    except (ValueError, TypeError):
        return "", 0, ""
    if not isinstance(payload, dict):
        return "", 0, ""
    code = payload.get("code")
    try:
        code = int(code) if code is not None else 0
    except (TypeError, ValueError):
        code = 0
    return str(payload.get("s") or ""), code, str(payload.get("message") or "")


def _refused(s: str, code: int) -> bool:
    """Whether the envelope says the call failed.

    ``s:"error"`` is the documented signal. A negative code with no ``s`` at
    all is treated the same way, because the token-failure responses seen in
    the wild carry the code and message but not always the status field.
    """
    if s.strip().lower() == "error":
        return True
    return not s and code < 0


def _classify(status: int, code: int, message: str) -> APIError:
    """Build the error class the status and envelope code call for.

    The envelope's code is consulted before the status, because FYERS reports
    both and the code is the more specific: a 400 carrying -8 is an expired
    token, not a bad request, and re-sending it would only be refused again.
    """
    if code in TOKEN_CODES or status == 401:
        return TokenExpiredError(status, code, message)
    if code == CODE_RATE_LIMITED or status == 429:
        return RateLimitedError(status, code, message)
    if status >= 500:
        return TransientError(status, code, message)
    return RejectedError(status, code, message)


class Transport:
    """Issues requests, throttles them, retries the ones worth retrying."""

    def __init__(
        self,
        *,
        auth_provider: Callable[[], str],
        base_url: str,
        data_url: str,
        client: httpx.Client | None = None,
        rate_per_sec: float = 8.0,
        burst: float = 8.0,
        timeout: float = 30.0,
        attempts: int = DEFAULT_ATTEMPTS,
        retry_base: float = DEFAULT_RETRY_BASE,
        sleep: Callable[[float], None] = time.sleep,
    ) -> None:
        """Build a transport.

        Args:
            auth_provider: Returns the current ``Authorization`` value --
                ``appId:accessToken``, or an empty string before login -- called
                per request so a token replaced mid-session is picked up.
            base_url: The trading root (``/api/v3``).
            data_url: The market-data root (``/data``). The split is FYERS's:
                quotes and history live on a different root from the account.
            client: Pre-built httpx client. Tests inject a MockTransport.
            rate_per_sec: Throttle refill rate. FYERS allows 10/s; the default
                sits under it.
            burst: Throttle bucket size.
            timeout: Per-request timeout, when no client is supplied.
            attempts: How many times a retryable refusal is sent. 1 disables
                retrying.
            retry_base: First backoff, doubled per attempt and overridden by a
                ``Retry-After`` header.
            sleep: Injected so tests do not actually wait.

        """
        self._auth_provider = auth_provider
        self.base_url = base_url.rstrip("/")
        self.data_url = data_url.rstrip("/")
        self._client = client or httpx.Client(timeout=timeout)
        self._bucket = TokenBucket(rate_per_sec, burst, sleep)
        self._attempts = max(1, attempts)
        self._retry_base = retry_base
        self._sleep = sleep

    def request(
        self,
        method: str,
        path: str,
        *,
        base: str | None = None,
        params: dict[str, Any] | None = None,
        json_body: dict[str, Any] | list[Any] | None = None,
        retry: bool = False,
    ) -> dict[str, Any]:
        """Issue a request and return the decoded JSON body.

        Args:
            method: HTTP method.
            path: Path below ``base``.
            base: API root, defaulting to the trading one.
            params: Query parameters.
            json_body: JSON request body.
            retry: Whether a transient refusal may be re-sent. False for every
                order mutation: retrying a POST that placed an order places a
                second one, and a transport error gives no way to tell which
                happened. It is explicit rather than derived from the method,
                because FYERS cancels an order with DELETE and modifies one
                with PATCH, and neither is safe to repeat either.

        Returns:
            The decoded response body.

        Raises:
            APIError: On a terminal refusal, or a retryable one that survives
                every attempt.

        """
        url = f"{(base or self.base_url).rstrip('/')}{path}"
        attempts = self._attempts if retry else 1
        last: APIError | None = None

        for attempt in range(1, attempts + 1):
            self._bucket.acquire()
            try:
                response = self._client.request(method, url, params=params, json=json_body, headers=self._headers())
            except httpx.HTTPError as exc:
                # A transport failure has no status to classify, and repeating
                # it is legitimate: the connection, not the request, failed.
                last = TransientError(0, 0, str(exc))
                if attempt == attempts:
                    raise last from exc
                self._sleep(self._backoff(None, attempt))
                continue

            body = response.text
            s, code, message = _envelope(body)
            if response.status_code < 400 and not _refused(s, code):
                try:
                    payload = response.json()
                except ValueError as exc:
                    # A body that will not decode is FYERS answering something
                    # other than what it documents. The same malformed answer
                    # will arrive again, so it is not retryable.
                    raise RejectedError(response.status_code, 0, f"decoding {path}: {exc}") from exc
                return payload if isinstance(payload, dict) else {}

            last = _classify(response.status_code, code, message or body[:300])
            if attempt == attempts or not last.retryable:
                raise last
            delay = self._backoff(response, attempt)
            log.warning(
                "fyers %s %s returned %s (code %s) - retrying in %.2fs (attempt %d/%d)",
                method,
                path,
                response.status_code,
                code,
                delay,
                attempt,
                attempts,
            )
            self._sleep(delay)

        assert last is not None
        raise last

    def _headers(self) -> dict[str, str]:
        """Build request headers with the current ``appId:token`` pair.

        The app id travels with every request. It is the one detail a caller
        porting from Kite or Upstox would get wrong: a bare bearer token is
        answered with "invalid token" (-15) rather than anything that names
        the missing part.
        """
        headers = {"accept": "application/json"}
        auth = self._auth_provider()
        if auth:
            headers["Authorization"] = auth
        return headers

    def _backoff(self, response: httpx.Response | None, attempt: int) -> float:
        """Seconds to wait before the next attempt.

        The server's own wait wins over the client's schedule, because it knows
        when its window resets and the client does not -- but it is capped, so
        an absurd value from a confused proxy cannot stall a sync for the rest
        of the day. FYERS sends both ``Retry-After`` in seconds and
        ``X-Retry-After-Ms``; the millisecond header is preferred because it is
        the precise one.
        """
        if response is not None:
            millis = response.headers.get("X-Retry-After-Ms")
            if millis:
                try:
                    ms = int(millis)
                except ValueError:
                    ms = 0
                if ms > 0:
                    return min(ms / 1000.0, MAX_RETRY_WAIT)
            header = response.headers.get("Retry-After")
            if header:
                try:
                    seconds = float(header)
                except ValueError:
                    seconds = 0.0
                if seconds > 0:
                    return min(seconds, MAX_RETRY_WAIT)
        return min(self._retry_base * (2 ** (attempt - 1)), MAX_RETRY_WAIT)

    def close(self) -> None:
        """Close the underlying HTTP client."""
        self._client.close()

File: go/fyers/transport.go
package fyers

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"time"

	"github.com/althk/tradekit/go/core/ratelimit"
)

// Bounds on a single call's retry behaviour.
//
// Four attempts with a two-second base doubles to 2s, 4s, 8s — about fourteen
// seconds of patience, which is long enough to ride out a per-second rate
// limit window and short enough that a stuck sync is noticed rather than
// hanging until the session ends. The per-minute window is a different matter:
// FYERS blocks an account for the rest of the day after the third breach, so
// the transport honours Retry-After rather than guessing.
const (
	defaultAttempts     = 4
	defaultRetryBase    = 2 * time.Second
	defaultMaxRetryWait = 60 * time.Second
)

// maxResponseBytes bounds a response read. The symbol master is fetched by a
// different path; nothing on the JSON API is legitimately this large, and an
// unbounded ReadAll against a confused endpoint is how a sync exhausts memory.
const maxResponseBytes = 16 << 20

// request is one call's parameters.
type request struct {
	method string
	// base overrides the client's API root. Market data lives on a
	// different root from trading, so the base cannot be a property of the
	// client alone.
	base    string
	path    string
	query   url.Values
	body    any
	out     any
	limiter *ratelimit.Limiter

	// retry allows the call to be repeated on a transient failure.
	//
	// It is explicit, and false for every order mutation, because retrying
	// a POST that placed an order places a second one. Making it a property
	// of the HTTP method instead would be wrong, since FYERS cancels an
	// order with DELETE and modifies one with PATCH, and neither of those
	// is safe to repeat either.
	retry bool
}

// envelope is the header every FYERS response carries. S is "ok" or "error";
// Code is FYERS's own status, which is not the HTTP status and is negative for
// every refusal.
type envelope struct {
	S       string `json:"s"`
	Code    int    `json:"code"`
	Message string `json:"message"`
}

// doJSON performs a request, decodes the JSON response into req.out, and
// classifies any failure into one of the three sentinels.
//
// Retries are bounded, honour Retry-After when the server sends one, and stop
// immediately on a cancelled context — a scan that has been cancelled must stop
// making requests, not work through its backlog.
func (c *Client) doJSON(ctx context.Context, req request) error {
	attempts := 1
	if req.retry {
		attempts = c.attempts
	}

	var last error
	for attempt := 1; attempt <= attempts; attempt++ {
		if err := ctx.Err(); err != nil {
			return err
		}
		lim := req.limiter
		if lim == nil {
			lim = c.general
		}
		if err := lim.Wait(ctx); err != nil {
			return err
		}

		retryAfter, err := c.doOnce(ctx, req)
		if err == nil {
			return nil
		}
		last = err

		var apiErr *APIError
		if !errors.As(err, &apiErr) || !apiErr.Retryable() || attempt == attempts {
			return err
		}
		if err := c.sleep(ctx, backoff(c.retryBase, attempt, retryAfter)); err != nil {
			return err
		}
	}
	return last
}

// doOnce issues one request. It returns the Retry-After the server asked for,
// in addition to the error, so the caller can honour it rather than guessing.
func (c *Client) doOnce(ctx context.Context, req request) (retryAfter time.Duration, err error) {
	base := req.base
	if base == "" {
		base = c.baseURL
	}
	u := base + req.path
	if len(req.query) > 0 {
		u += "?" + req.query.Encode()
	}

	var reader io.Reader
	if req.body != nil {
		raw, err := json.Marshal(req.body)
		if err != nil {
			return 0, fmt.Errorf("fyers: encoding request body for %s: %w", req.path, err)
		}
		reader = bytes.NewReader(raw)
	}

	httpReq, err := http.NewRequestWithContext(ctx, req.method, u, reader)
	if err != nil {
		return 0, fmt.Errorf("fyers: building request %s %s: %w", req.method, req.path, err)
	}
	httpReq.Header.Set("Accept", "application/json")
	if h := c.authHeader(); h != "" {
		httpReq.Header.Set("Authorization", h)
	}
	if req.body != nil {
		httpReq.Header.Set("Content-Type", "application/json")
	}

	resp, err := c.http.Do(httpReq)
	if err != nil {
		// A transport failure has no status to classify, and repeating it
		// is legitimate: the connection, not the request, is what failed.
		return 0, &APIError{Message: err.Error(), sentinel: ErrTransient}
	}
	defer func() { _ = resp.Body.Close() }()

	raw, err := io.ReadAll(io.LimitReader(resp.Body, maxResponseBytes))
	if err != nil {
		return 0, &APIError{StatusCode: resp.StatusCode, Message: err.Error(), sentinel: ErrTransient}
	}
	retryAfter = parseRetryAfter(resp.Header.Get("Retry-After"), resp.Header.Get("X-Retry-After-Ms"))

	env := readEnvelope(raw)
	// A refusal can arrive under a 2xx: the reference documents an order
	// rejection as HTTP 200 with s:"error" and code -99, and a token
	// failure on some endpoints the same way. The body is therefore
	// inspected even when the status says the call succeeded.
	if resp.StatusCode >= 400 || env.refused() {
		message := env.Message
		if message == "" {
			message = truncate(strings.TrimSpace(string(raw)), 300)
		}
		return retryAfter, &APIError{
			StatusCode: resp.StatusCode,
			Code:       env.Code,
			Message:    message,
			sentinel:   classify(resp.StatusCode, env.Code),
		}
	}

	if req.out != nil {
		if err := json.Unmarshal(raw, req.out); err != nil {
			// A body that will not decode is FYERS answering something
			// other than what it documents. It is not retryable: the
			// same malformed answer will arrive again.
			return 0, &APIError{
				StatusCode: resp.StatusCode,
				Message:    fmt.Sprintf("decoding %s: %v", req.path, err),
				sentinel:   ErrRejected,
			}
		}
	}
	return 0, nil
}

// refused reports whether the envelope says the call failed.
//
// s:"error" is the documented signal. A negative code with no s at all is
// treated the same way, because the token-failure responses seen in the wild
// carry the code and message but not always the status field.
func (e envelope) refused() bool {
	if strings.EqualFold(strings.TrimSpace(e.S), responseError) {
		return true
	}
	return e.S == "" && e.Code < 0
}

// readEnvelope decodes the header fields, tolerating a body that is not an
// object at all — the history endpoint on an empty range has been seen to
// answer with a bare array — by returning an empty envelope.
func readEnvelope(raw []byte) envelope {
	var env envelope
	_ = json.Unmarshal(raw, &env)
	return env
}

// backoff returns how long to wait before an attempt, preferring the server's
// own Retry-After over the client's exponential schedule.
//
// The server knows when its window resets and the client does not, so a
// Retry-After is obeyed even when it is longer than the schedule would be — but
// it is capped, because an absurd value from a confused proxy would otherwise
// stall a sync for the rest of the day.
func backoff(base time.Duration, attempt int, retryAfter time.Duration) time.Duration {
	if retryAfter > 0 {
		if retryAfter > defaultMaxRetryWait {
			return defaultMaxRetryWait
		}
		return retryAfter
	}
	d := base << (attempt - 1)
	if d > defaultMaxRetryWait {
		return defaultMaxRetryWait
	}
	return d
}

// parseRetryAfter reads the wait the server asked for.
//
// FYERS sends both the standard Retry-After in seconds and X-Retry-After-Ms
// in milliseconds. The millisecond header is preferred when present because
// it is the precise one: a Retry-After of "1" for a 150ms window would make a
// ten-order burst take ten seconds instead of one and a half.
func parseRetryAfter(seconds, millis string) time.Duration {
	if ms, err := strconv.ParseInt(strings.TrimSpace(millis), 10, 64); err == nil && ms > 0 {
		return time.Duration(ms) * time.Millisecond
	}
	seconds = strings.TrimSpace(seconds)
	if seconds == "" {
		return 0
	}
	if secs, err := strconv.ParseFloat(seconds, 64); err == nil {
		if secs <= 0 {
			return 0
		}
		return time.Duration(secs * float64(time.Second))
	}
	if at, err := http.ParseTime(seconds); err == nil {
		if d := time.Until(at); d > 0 {
			return d
		}
	}
	return 0
}

// truncate bounds a message taken from a response body, so one confused
// endpoint cannot fill the log with megabytes of HTML.
func truncate(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[:n] + "..."
}
